Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+            // space and shuffle that many bytes around, it's likely no faster
+            // than reallocating and copying.
+            var newBytes = self._storage
+            newBytes.clear()
+            newBytes.write(bytes: self._storage.viewBytes(at: r, length: self._storage.capacity - r))
+            newBytes.write(bytes: self._storage.viewBytes(at: 0, length: w))
+            self._storage = newBytes
+            self._moveHead(to: 0)
+            self._moveTail(to: _storage.capacity - r + w)
+        }
+    }
+    
+    @inlinable
+    mutating func _moveHead(to newIndex: Int) {
+        assert(newIndex >= 0 && newIndex <= self._storage.capacity)
+        self._ringHead = newIndex
+    }
+    
+    @inlinable
+    mutating func _moveHead(forwardBy amount: Int) {
+        assert(amount >= 0)
+        let readerIsBelowWriter = self._ringHead <= self._ringTail
+        var newIndex = self._ringHead + amount
+        
+        // reader index should not pass writer index
+        if readerIsBelowWriter {
+            assert(newIndex <= self._ringTail)
+        } else if newIndex >= self._storage.capacity {
+            // wrap it around
+            newIndex -= self._storage.capacity
+            // now test that it didn't pass writer
+            assert(newIndex <= self._ringTail)
+        }
+        
+        self._moveHead(to: newIndex)
+        self._readableBytes -= amount
+    }
+    
+    @inlinable
+    mutating func _moveTail(to newIndex: Int) {
+        assert(newIndex >= 0 && newIndex <= self._storage.capacity)
+        self._ringTail = newIndex
+    }
+    
+    @inlinable
+    mutating func _moveTail(forwardBy amount: Int) {
+        assert(amount >= 0)
+        let writerIsBelowReader = self._ringTail < self._ringHead
+        var newIndex = self._ringTail + amount
+        
+        // writer index should not pass readerIndex
+        if writerIsBelowReader {
+            assert(newIndex <= self._ringHead)
+        } else if newIndex >= self._storage.capacity {
+            // wrap it around
+            newIndex -= self._storage.capacity
+            // now test it
+            assert(newIndex <= self._ringHead)
+        }
+        
+        self._moveTail(to: newIndex)
+        self._readableBytes += amount
+    }
+    
+    @inlinable
+    mutating func moveHead(forwardBy amount: Int) {
+        self._moveHead(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func moveTail(forwardBy amount: Int) {
+        self._moveTail(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func unwrite(byteCount: Int) {
+        assert(byteCount >= 0)
+        if byteCount <= self._ringTail {
+            self._ringTail -= byteCount
+        } else {
+            self._ringTail = self._storage.capacity - byteCount + self._ringTail
+        }
+        self._readableBytes -= byteCount
+    }
+    
+    @inlinable
+    mutating func unread(byteCount: Int) {
```

Comment:
Is this method actually used? If it is, it needs to be fixed up as the above, by bumping asserts to preconditions and verifying we don't break the ring invariant.